### ISSUE TITLE
Pro Tools "AAE error -14034" - GetControlWithTag() returning null pointer from "valid" UI

### ISSUE THREAD

**Issue by rexmartin158** [iPlug2/iPlug2#746]
**Describe the bug**
AAX plugin loaded in Pro Tools on Windows throws an "AAE error -14034".  Closing and re-opening the plugin UI thereafter works normally but removing and re-loading the plugin again throws the error.

![image](https://user-images.githubusercontent.com/84396036/118905139-04e8b500-b8d0-11eb-8e31-68d536f9a92e.png)

**To Reproduce**
Debugger shows error is thrown by a null pointer from GetControlWithTag() in OnParamChangeUI() when plugin is loaded:

```
void TestPlug::OnParamChangeUI(int paramIdx, EParamSource source)
{
    if(GetUI())
    {
        if(paramIdx == kProcessControl) GetUI()->GetControlWithTag(kCtrlTagPowerLED)->SetValueFromDelegate(Process);
    }
}
```

![image](https://user-images.githubusercontent.com/84396036/118903114-dec11600-b8cb-11eb-99b4-88f80a8e92b3.png)

**Expected behaviour**
GetControlWithTag() should return a valid pointer and not a null pointer when "if(GetUI())" is true.  Seems "if(GetUI())" check may be returning the wrong value whereby UI does not exist at the time of the GetControlWithTag() call.

Commenting out the above code eliminates the AAE -14034 error.  Also works properly if control update code is moved to OnIdle() - (but repeated updates in OnIdle() for such a control is not needed or desired/efficient).

**IMPORTANT DETAILS**
* What plug-in format does it relate to: have only seen this happen in AAX/Pro Tools on Windows - no debugger error launching VST3 on Reaper
* What platform does it relate to: have only seen this happen in Pro Tools on Windows - Pro Tools on Mac does not throw this error

**Comment by rexmartin158** [comment]
Upon further investigation it appears this problem is only happening with "GetControlWithTag()" i.e., controls not attached to a parameter.  Updating a control that _**is**_ attached to a parameter - for example with "GetUI()->DisableControl()" - does not throw the error.  

I have worked around the issue by adding a null pointer check to the GetControlWithTag() update like this:

```
void TestPlug::OnParamChangeUI(int paramIdx, EParamSource source)
{
    if(GetUI())
    {
        if (paramIdx == kProcessControl && GetUI()->GetControlWithTag(kCtrlTagPowerLED))
        {
            GetUI()->GetControlWithTag(kCtrlTagPowerLED)->SetValueFromDelegate(Process);
        }
    }
}
```
but I don't understand why that null ptr check is necessary if "if(GetUI()" is true.  Will try to see what Pro Tools is doing during plugin construct.  I suspect Pro Tools is calling OnParamChange() before the UI is constructed - but if so "if(GetUI()" should return false.  So...IDK

**Comment by TBProAudio** [comment]
It is fairly clear, that OnParamChange(UI) could be called before GetControlWithTag can deliver valid information.
Usually "kCtrlTagPowerLED" is attached to the control with AttachControl() ...

**Comment by rexmartin158** [comment]
> Usually "kCtrlTagPowerLED" is attached to the control with AttachControl().
> So AttachControl (withTag) should come before GetControlWithTag() !

Not sure what you're saying here.  I have it set up correctly, AFAIK.  The control "kCtrlTagPowerLED" is initiated in mLayoutFunc in the plugin constructor along with all my other controls.

```
  mLayoutFunc = [&](IGraphics* pGraphics)
  {
....
    pGraphics->AttachControl(new IBitmapControl(228, 97, LED, -1), kCtrlTagPowerLED);//---------Power LED
....
}:
```

If GetUI() returns "true" when checked in OnParamChangeUI() the control should exist, right?  Why is it returning a null pointer when the plugin first opens - and why only in Pro Tools?  Debugger shows the control is created before OnParamChangeUI() is called - so I don't know what's causing this unless if(GetUI()) is returning true when it should be false.

**Comment by TBProAudio** [comment]
OK, we never use OnParamChangeUI, only OnParamChange.
Need to check OnParamChangeUI.

**Comment by AlexHarker** [comment]
I can't quite see how this could happen but I might be missing something.

IGraphics gets made in IGraphicsEditorDelegate::OpenWindow, which also calls through to IGraphics::OpenWindow() which then lays out the UI (via IGraphicsEditorDelegate::LayoutUI()). If the UI is opening for the first time nothing happens in between (otherwise the UI may resize to the last size when closed). Methods are marked final so you can't be overriding IGEditorDelegate::OpenWindow()...

I can't see presently how GetUI() can return true but there are no controls unless there's some multithreading going on - I think we'd need more debugging info - probably breaking in IGEditorDelegate::OpenWindow on construction to see where you are in this method when it crashes.

**Comment by rexmartin158** [comment]
> I can't see presently how GetUI() can return true but there are no controls unless there's some multithreading going on - I think we'd need more debugging info - probably breaking in IGEditorDelegate::OpenWindow on construction to see where you are in this method when it crashes.

Not sure how to dig deeper but the debugging info I provided above at least points out the root of the problem - during construction, OnParamChangeUI() is being called - and if(GetUI()) is returning "true" - when in fact the UI hasn't been created yet.  This only occurs at construction.  When the UI is subsequently closed/re-opened it works properly.

Knowing this has allowed me to create several work-arounds and they all fix the problem.  One method is to check for a valid CtrlTag pointer before attempting to update that control (shown in one of my posts above).  Another successful method - even simpler/more broad - is to check if mLayoutFunc has completed by setting a bool variable at the very end of that function and checking if true in OnParamChangeUI().  That approach could, perhaps, be built-in to the call to OnParamChangeUI() which would eliminate a need for any "if(GETUI()" checks in OnParamChangeUI().

**Comment by rexmartin158** [comment]
> OK, we never use OnParamChangeUI, only OnParamChange.
> Need to check OnParamChangeUI.

I initially updated all of my Tagged controls in OnIdle() but then realized I was constantly updating stuff that didn't need repeated updating.  So I have now moved all such control updates to OnParamChangeUI() where they are updated only once and only if required due to a parameter change.  I believe that is the intention of OnParamChangeUI().

**Comment by olilarkin** [comment]
Controls that are linked to parameters shouldn't need manual updating. FWIW the "UI" in OnParamChangeUI means UI/main thread. OnParamChangeUI() that is guaranteed on the main thread, where as OnParamChange() could happen on the audio thread.

**Comment by rexmartin158** [comment]
> Controls that are linked to parameters shouldn't need manual updating. 

The issue is with controls that are not attached to parameters.   "if(GetUI()" returns "true" when they haven't been created yet - confirmed by a null pointer to the control - which throws an error in Pro Tools.

**Comment by olilarkin** [comment]
Could you modify an example e.g.  IPlugEffect on a branch so that it crashes and I'll take a look?

**Comment by AlexHarker** [comment]
Agreed that a simple example would be really helpful here - I'm not setup to run pro tools here, so I can only check by reading the source.

> if(GetUI()) is returning "true" - when in fact the UI hasn't been created yet

That's the bit I can't understand - the creation of IGraphics and the layout should be basically sequential, so I can't see how the host would be able to call OnParamChangeUI in the middle of that, but as I stated above I might be missing something...

**Comment by rexmartin158** [comment]
>>but it's doesn't seem like a simple construction ordering error, otherwise we'd see it on other formats.

It **_is_** in other formats.  I have since checked VST3 and it is there as well.  I did not check further.

To be clear this issue does not "crash" the DAW - but in Pro Tools it throws an AAE -14034 error (as I show in the OP here) which appears as a broken plugin.

Debugger shows this as a "read access violation error" - so it is a real problem - but some DAWs - like Reaper - seem to work past it.  Pro Tools throws an error but continues to run upon closing the error window.  


>>Agreed that a simple example would be really helpful here

This is easy to check using any project that employs OnParamChangeUI().  Set a breakpoint at OnParamChangeUI() and another one at mLayoutFunc (where the controls are created).  Load the plugin on a track in a DAW and you will see that OnParamChangeUI() gets called **_before_** mLayoutFunc.  For some reason, checking if(GetUI() in OnParamChangeUI() at that point is returning "true" when in fact the controls haven't been created yet.  This is only happening during plugin construction.  Afterwards (when closing and re-opening the plugin UI) if(GetUI() reports correctly.

Besides the return value of if(Get()UI), what I am secondarily suggesting is that the function "OnParamChangeUI()" itself should perhaps not be called/allowed to run unless there is a fully created UI.  It appears it is currently being triggered every time OnParamChange() is called which creates the "update control vs create control" sequencing problem during plugin construction.

I have fixed the issue myself with the 2 different solutions I posted above.  Both means check to see if the controls exist before attempting to update them.  Simply checking if(GetUI() - by itself - is not always giving the right answer.

**Comment by AlexHarker** [comment]
> This is easy to check using any project that employs OnParamChangeUI(). Set a breakpoint at OnParamChangeUI() and another one at mLayoutFunc (where the controls are created). Load the plugin on a track in a DAW and you will see that OnParamChangeUI() gets called **_before_** mLayoutFunc.

I've just checked this and perhaps I need to be clearer. This is what I see under VST3:
1 - the plugin constructs
2 - OnParamReset is called which causes multiple calls to OnParamChangeUI()
3 - OpenWindow() is called to create the UI
4 - LayoutUI() is called (which calls mLayoutFunc if there is one)

**That is all fine and correct** - the UI doesn't exist at step 2 so there is no issue here.

The issue would arise if we saw calls to OnParamChangeUI() in between steps 3 and 4 (as above you say you are getting calls to OnParamChangeUI() before the GUI has fully been made but after it has started to be made (so GetUI() is returning true). On my previous readings of the code I cannot yet see how that is possible and that's the breakpoints that we need to confirm (a break in OpenWindow() then OnParamChangeUI() and then LayoutUI() in that order). If we can't confirm that order of class then we don't yet understand what the issue is in full and therefore can't fix it more properly or determine if it is something specific to your plugin code.

**Comment by sstillwell** [comment]
I have run into similar issues where OnIdle() can get called before the UI is completely built.  My workaround has been to define a member variable mAllowUpdates defaulted to false in the constructor, then set mAllowUpdates = true at the end of mLayoutFunc.  in OnIdle() I use
if (this->GetUI() && mAllowUpdates){ } bracketed around anything that might touch the UI (meter updates, in my case).  Maybe a bit of a kluge, but it works.

**Comment by AlexHarker** [comment]
Thanks @sstillwell, but again I find this confusing - the UI is created on the main thread and the timer is also called on the main thread - therefore it shouldn't be possible for the timer to interrupt the UI creation process. This is also not something I've seen. Surreal machines' Impact, for instance, which I worked on calls meters in OnIdle() and I would expect if this were a general issue for a plugin with a fairly involved UI like this which is making using of OnParamChangeUI() and OnIdel() to exhibit the issue.

If either of you has an example plug where removing your additional checks creates the issue then I'd be happy to compile it to investigate further.

**Comment by sstillwell** [comment]
I don't have an example that I can readily give you, but literally the only things in my OnIdle() method are calls to some IBufferSender's TransmitData() methods (which in turn gets their ProcessBlock() methods called on the audio thread in the plugin's ProcessBlock() method.  Do you have EnableMouseOver(true) set in the Layout function?  That's the only slightly non-default thing I'm doing that I can think of...

**Comment by AlexHarker** [comment]
I do have EnableMouseOver(true) set. Any further info on platform/plugin formats would be gratefully received. Stack traces for the problematic calls would be super helpful (what is the full sequence of calls) if you can generate those (I understand there might be complexities to sharing some kinds of code)

At the moment there are only two ways I could imagine the sort of problems outlined above:

1 - something in the Layout function/elsewhere is calling OnParamChangeUI() (which is probably a bad idea, but perhaps we should guard against it, as I could imagine ways in which this might occur unintentionally -  I could most probably fix that.

2 - there is a threading bug somewhere (OpenWindow() is not threadsafe and these calls are being made from the wrong threads).

Those are both more likely for the OnParamChangeUI() bug rather than OnIdle() but the first thing to fixing these issues is to get a reproducible example, because there doesn't seem to be anything obvious I can reason through from the code alone.

**Comment by rexmartin158** [comment]
@AlexHarker 

I do not have an example I can share here at the moment but I traced this down (and developed a workaround for it) as I posted above.  I know what I'm reporting is confusing so I will try again here.

The issue is with controls that are not attached to parameters. The values of these controls are set using  GetControlWithTag()->SetValueFromDelegate(), GetUI()->DisableControl(), etc. in either OnParamChangeUI() or in OnIdle().  Often times these display controls may correlate with parameter settings in which case it is most efficient to set them in OnParamChangeUI() rather than constantly updating them in OnIdle().

The problem arises when we try to set or update these controls before they exist which throws the AAE/"null pointer" error in Pro Tools (and will likely crash other DAWs at some point too).

Here is the sequence I have observed in debugger in Pro Tools:

A) When the plugin is first loaded, OnParamChangeUI() gets called before mLayoutFunc.  At this point, any calls to GetControlWithTag()) in OnParamChangeUI() will throw the "null pointer" error in Pro Tools because the control hasn't been created yet.  Checking for existence of the controls with "if(GetUI())" doesn't work as sometimes it returns "true" when in fact the controls with tags do not yet exist.

OnParamChangeUI() is then called a second time, after mLayoutFunc, at which point controls with tags _can_ be updated.

B) When an established plugin UI is subsequently closed then re-opened, OnParamChangeUI() gets called **_after_** mLayoutFunc - at which point the controls and UI have been created - and GetControlWithTag() works as expected.


My workaround (posted above) sets a bool variable I named "controlsready" to "true" at the end of the mLayoutFunc which I then check in OnParamChangeUI() along with if (GetUI()) before update/setting any controls with tags.  This workaround has eliminated the AAE error in Pro Tools.  

The formal fix, IMO, would be to prevent OnParamChangeUI() from running before all controls have been created by waiting until mLayoutFunc is complete (as it currently works on UI close/reopen).

**Comment by AlexHarker** [comment]
Hi @rexmartin158 - thanks for taking the time to go through it, but I'm not confused about what you are seeing I have followed all of that - what I am struggling with is *how it happens that you call GetUI() and it returns incorrectly*.  Anything else is to my mind fixing the symptoms, not the underlying problem.

OnParamChangeUI() will currently always be called before the Layout function is called - that should be safe, and when I observe it on VST3 (as above) things occur in the right order (GetUI() is correct and there is no unexpected behaviour). The fix will likely *not* be to stop calling OnParamChangeUI() before the layout function (if I was going to do that I'd add a helper that checks GetUI() whenever that is called which would prevent users having to write it themselves, but that's a separate issue as *if GetUI() returns incorrectly then that wouldn't work to fix this bug*). The fix to your problem requires us to know *why* GetUI() returns incorrectly. 

Sadly, the workaround doesn't help me understand that - it just stops you dereferencing a null pointer (which is not safe to do).

To understand why GetUI() returns incorrectly I would like to know the full stack trace that you get when Pro Tools gives the error (all the method calls in the stack that come *before* the OnParamChangeUI() call that throws the error (to do this you would remove your workaround, and set an set a test for a nullptr on the relevant line and then break on that). Then copy out the stack trace here - I would hope that would give me a much clearer picture - as I don't have a plugin that is doing this that's not something I can do myself at present.

**Comment by rexmartin158** [comment]
@AlexHarker 

Thank you for bearing with me. 

It's going to take me some time to put together a demo, sign it with Pace and all that stuff.  Will try to post something here by tomorrow.

But yes, it seems the root problem is that GetUI() sometimes returns "true" when all controls/UI have not yet been created.  Everything inside my OnParamChangeUI() code block is wrapped with if(GetUI()) yet that fails to prevent the null pointer/AAE error as shown at the top of this thread.

**Comment by sstillwell** [comment]
The issue isn't that GetUI() is returning NULL or not so much as it is that all of the controls on the UI may not yet be created, and any pointers that you might capture to reference them within the layout func (or by trying to find them by TagId) may give you an invalid reference since the specific control you're looking for may not be created yet.

It seems that the base IGraphics pointer pGraphics is getting created and is valid, but the LayoutFunc has not yet been called at the time OnParamChangeUI gets called the first time (or in my case, during OnIdle()) - GetUI() returns a valid pointer to the graphics object, but the individual CONTROL objects may not be valid yet.  Checking GetUI() for validity is necessary but not sufficient if you're going to be manipulating individual control elements owned by the UI.

Alex, Rex, sorry if I'm inserting irrelevant comments here - I'll stop if I'm just decreasing the signal-to-noise ratio.  Let me know.

**Comment by AlexHarker** [comment]
In general you probably shouldn't keep pointers to controls any more ini plug2 - you should use tags IMO.

Agreed @sstillwell that GetUI() should be sufficient, but isn't - the thing is that breaks some fundamental assumption that we should be able to rely on about threading and ordering (timer is on main thread / GUI is on main thread / opening of GUI and layout are sequential and shouldn't call back to OnParamChangeUI() until done). I'd like to get to the bottom of why those assumptions are wrong and fix accordingly, rather than add workarounds ontop.

FWIW I can see how OnParamChangeUI() *could* get called in the middle of the Layout function, but it would need to be caused by some custom control code, and it'd be useful to know if that is the case, and if so why. I don't understand at all why OnIdle()/OnTimer() would be able to call at that point (unless someone was explicitly calling them).

Are you both primary testing on windows? And if so what graphics backends? That might provide a clue that could help and I can try to match your setups with a more complex plug to see if I can repro (I test mostly on Mac and might have chosen graphics options which change behaviour under windows).

**Comment by AlexHarker** [comment]
Also @sstillwell - if I can get a full trace for the call that crashes/would crash without the additional check that would help a lot.

**Comment by sstillwell** [comment]
I'm developing and testing primarily on macOS, with Windows testing after I feel I've got a stable build.  We're using Skia w/Metal on macOS, Skia CPU on Windows.  IPlug2 current at [ab2726c49]. I'll see if I can get a stack trace.

**Comment by rexmartin158** [comment]
@AlexHarker 

I removed my workaround code from one of my AAX builds and it again fails when the plugin is first loaded, throwing the AAE error as shown at the top of this thread.  It fails at OnParamChangeUI() due to null pointer in call to the first GetControlWithTag().

Here is the call stack up to that point:

>	dBSmooth.aaxplugin!dBSmooth::OnParamChangeUI(int paramIdx, iplug::EParamSource source) Line 1297	C++
 	dBSmooth.aaxplugin!iplug::IEditorDelegate::SendParameterValueFromDelegate(int paramIdx, double value, bool normalized) Line 225	C++
 	dBSmooth.aaxplugin!iplug::igraphics::IGEditorDelegate::SendParameterValueFromDelegate(int paramIdx, double value, bool normalized) Line 124	C++
 	dBSmooth.aaxplugin!iplug::IEditorDelegate::SendCurrentParamValuesFromDelegate() Line 171	C++
 	dBSmooth.aaxplugin!dBSmooth::OnParamChange(int paramIdx) Line 915	C++
 	dBSmooth.aaxplugin!iplug::IEditorDelegate::OnParamChange(int paramIdx, iplug::EParamSource source, int sampleOffset) Line 113	C++
 	dBSmooth.aaxplugin!iplug::IPlugAAX::UpdateParameterNormalizedValue(const char * paramID, double iValue, AAX_EUpdateSource iSource) Line 249	C++
 	[External Code]	
 	dBSmooth.aaxplugin!AAX_VAutomationDelegate::PostSetValueRequest(const char * iParameterID, double iNormalizedValue) Line 65	C++
 	dBSmooth.aaxplugin!AAX_CParameter<double>::SetValue(double newValue) Line 643	C++
 	dBSmooth.aaxplugin!AAX_CParameter<double>::SetNormalizedValue(double normalizedNewValue) Line 925	C++
 	dBSmooth.aaxplugin!AAX_CEffectParameters::SetParameterNormalizedValue(const char * iParameterID, double aValue) Line 408	C++
 	dBSmooth.aaxplugin!iplug::IPlugAAX::SetChunk(unsigned int chunkID, const AAX_SPlugInChunk * pChunk) Line 507	C++
 	[External Code]	

This is on Windows 10 but I get the same error on Mac.   All of my Controls with Tags are IBitmapControls.  Nothing "custom".

As mentioned above, all of the code in my OnParamChangeUI() block is wrapped with "if GetUI()" which does not appear to be returning the correct value.  The code inside that test is running even with "null" control pointers (not yet created).

**Comment by AlexHarker** [comment]
Thanks @rexmartin158 - this is a really good start and it might be:

A - AAX specific
B - specific to what you do in OnParamChange()

The things that would be good to know to start with are:

1 - What thread is this (specifically setChunk() called on - I see a note to say. "// TODO: UI thread only?" and I suspect that assumption might be being violated.

2 - Let's try to test theory B - can you either comment out all of the code in OnParamChange() and see if it still crashes, or possibly post the code here?

It looks like your OnParamChange code calls into SendCurrentParamValuesFromDelegate() which it probably shouldn't do - that would send all the values from the processor to the UI, but if OnParamChange() is being called you should be expecting a call to OnParamChangeUI(). In general if you are calling things like Send....() methods you probably aren't doing it in an optimal way - iplug2 should remove the need to send params around explicitly in comparison to iplug1. Ideally we'd lock out various methods from the final class, but that's a bit hard to do and in some cases there are rare occasions you might need them.

**Comment by AlexHarker** [comment]
If you try 2 and it still crashes we might get a different stack trace, also...

**Comment by rexmartin158** [comment]
> If you try 2 and it still crashes we might get a different stack trace, also...

@AlexHarker 

Thank you, Alex.  It's 2) and may not be a bug after all.

Somewhere along the line I started using the "two parameter" version of OnParamChangeUI()  - with int paramIdx, iplug::EParamSource source.  I changed back to the single parameter version - that only uses int paramIdx - and now it works properly (no AAE error) without my "fix" workaround code.

So it seems I created my own problem trying to fix something else somewhere else.  I do not recall why I changed to the 2 param version of OnParamChangeUI() but that seems to be where the problem lies.  Either there is an issue with that overload or I used it when I shouldn't have.  My greatest apologies if I have wasted your time here.

**Comment by AlexHarker** [comment]
Hi @rexmartin158.

Thanks for looking into this further. I'm not sure I totally follow the issue with the version of OnParamChangeUI() you override. The default for the two parameter version is to call the one parameter version, but you should be able to override either version without issue.

The issue I saw was that your OnParamChange() (without UI) seemed to call through to OnParamChangeUI() indirectly, which doesn't seem right. It may be that you are just explaining some other complexities having reviewed your code, but it would be good to clarify that you've removed the callthrough from OnParamChange() to fix this.

No need to apologise for reporting a crash - hopefully we can fully clear up how it is happening and then I can look and see if there's anything we can do (given that info) to improve things further.

**Comment by rexmartin158** [comment]
> The issue I saw was that your OnParamChange() (without UI) seemed to call through to OnParamChangeUI() indirectly, which doesn't seem right.

Not sure what you're thinking here.  I don't call "OnParamChangeUI()" from anywhere in my code.  If something's running out of order I don't think it's from anything I'm doing.  Debugger shows OnParamChangeUI() runs automatically following every OnParamChange() - so I'm not sure what you mean by "indirectly".

I guess something that could still be looked into as a potential for trouble is what Stillwell pointed out:

> the LayoutFunc has not yet been called at the time OnParamChangeUI gets called the first time

That sequence of events - on initial plugin construction - seems like it could still cause problems because we could be attempting to set/update controls in OnParamChangeUI that haven't been created yet.  Checking "GetUI()" at that point is unreliable as it sometimes returns "true" when it is _not_ true - so perhaps there is something not quite right in GetUI() ?

I observed in debugger that when an established plugin UI is closed and then re-opened OnParamChangeUI doesn't run until **_after_** mLayoutFunc - which is where I got the idea for my hack.

**Comment by AlexHarker** [comment]
I outlined above that your stack trace seems to show a call to SendCurrentParamValuesFromDelegate() from your implementation of OnParamChange() (on line 915 of your code if I am reading the trace correctly) - that will call through to OnParamChangeUI() and I am not sure why you would be calling it at all - it doesn't seem like a good idea in general and I can't see why it would ever make sense in OnParamChange() but then I don't know what's happening in the surrounding code.

**Comment by AlexHarker** [comment]
( so by indirectly I mean you call a method that calls OnParamChangeUI() )

**Comment by rexmartin158** [comment]
> your stack trace seems to show a call to SendCurrentParamValuesFromDelegate() from your implementation of OnParamChange()

> I can't see why it would ever make sense in OnParamChange()

Some of my plugins have interactive controls that cannot be directly linked (there are math relationships between them).  I use SendParameterValueFromUI to update the associated parameter(s) and then SendCurrentParamValuesFromDelegate to update the UI controls.  AFAIK, there is no one function that does all of that in one step.  It seems to work just fine but if there is better way please advise.

> the LayoutFunc has not yet been called at the time OnParamChangeUI gets called the first time

This is the order of events I see when the plugin is constructed and it is still curious to me.  The very first run of OnParamChangeUI - who's purpose, AFAIK, is to update controls that do not have parameters - occurs before mLayoutFunc where the UI controls are created.  The safety check "if(GetUI())" returns "true" at that point yet all UI (the controls) haven't been created yet.  So how does this work in _any_ DAW?  It seems it shouldn't.  My hack - that prevented updating controls until mLayoutFunc completes - stopped the error messages in Pro Tools - so I thought I understood all of this (but obviously I don't).

Time for a break perhaps.  It's all working fine here now.  Again my apologies for using so much of your time on this.  Unless @sstillwell has more input I think this particular bug report can be closed as "resolved/user error".

**Comment by AlexHarker** [comment]
@rexmartin158 - if you have controls that are dependent on multiple parameters then they should declare themselves as such (each control can reference more than one param - see the IVXYPadControl control for instance) and do their maths internally. I cannot comment in detail on your design in detail without seeing the code, but the basic principle in iplug2 is that if you are manually sending parameters around it is unlikely that you are making best use of iplug2 and that approach is non-optimal,  prone to errors and not inline with the style of the examples (there is another matter about how to document this - particularly in relation to iplug1 where such things were often necessary, but there was significant work in iplug2 to simplify and avoid this for developers). If you have links between parameters or similar this might be different but it's impossible to advise without a clearer idea of what it is you are doing. However, I cannot imagine a situation in which SendCurrentParamValuesFromDelegate()  from in OnParamChange() would ever be correct. I'd be happy to advise on how to approach this if you can explain how your controls should work.

"regardless of when OnParamChangeUI is called..." - on this we disagree - you seem to insist that whenever *any* plugin constructs that the first call to OnParamChangeUI() the UI will not be fully constructed, but I do not believe this to be the case at all  I believe that the most likely explanation of your issue (which may be distinct from @sstillwell's as that relates to OnIdle() which yours does not), is that calls that you make in your plugin code lead to the issue, because otherwise I would expect to see it elsewhere (I have plugins with tagged controls and no additional checks that I have not seen exhibit this behaviour).

My suggestion above was to comment out everything in OnParamChange() *no UI* and remove your workaround. If the plugin stops crashing at that point then we can be sure that the issue is a result of calls you are making in OnParamChange() and is not a general/consistent iplug2 bug. I am not clear on whether you have tried this or not from the conversation above. Once that has been established we know whether or not the issue is a result of the calls you are making, or whether to continue looking for a more generic iplug2 issue. We can also then figure out whether or not to guard against this (even though it is not optimal design) or to consider this a developer issue. You also can then figure out a more optimal design.

**Comment by rexmartin158** [comment]
> My suggestion above was to comment out everything in OnParamChange() (_no UI_) and remove your workaround. If the plugin stops crashing at that point then we can be sure that the issue is a result of calls you are making in OnParamChange() and is not a general/consistent iplug2 bug. I am not clear on whether you have tried this or not from the conversation above.

Commenting out everything in my OnParamChange _**will**_ stop the error because OnParamChangeUI will not be called in that case - and that is where the problem is. 

OnParamChangeUI does not run automatically.  Unless I call it in my code - either directly or indirectly - it does nothing/never gets called by anything else.  It has to be called "manually" from somewhere and OnParamChange seems the logical place.  If OnParamChange is the wrong place to call it then where is the right place to call it?

> you seem to insist that whenever _any_ plugin constructs that the first call to OnParamChangeUI() the UI will not be fully constructed

It's not my opinion - it's what debugging in Pro Tools shows and also what you said further up in this thread:

> OnParamChangeUI() will currently always be called before the Layout function is called

So it is curious to me that this construction order works anywhere.  Until mLayoutFunc is complete, pointers to controls with tags return NULL because they don't exist yet.  Meanwhile, the check if(GetUI()) returns "true" and the code in OnParamChangeUI tries to run.  I posted a screenshot from debugger above that shows that result:

![image](https://user-images.githubusercontent.com/84396036/145336121-7168ae29-71dd-4f1b-bc0e-6e75c54697cc.png)


> ..that should be safe, and when I observe it on VST3 (as above) things occur in the right order (GetUI() is correct and there is no unexpected behaviour).

Have you tested it in _**Pro Tools**_?  That is where I discovered this problem.  I have not seen this issue in DAWs running other formats (VST, VST3 or AU). 

> I cannot imagine a situation in which SendCurrentParamValuesFromDelegate() from in OnParamChange() would ever be correct.

I don't need to imagine where it does or does not work.  I have tried calling SendCurrentParamValuesFromDelegate() in other parts of the code and it did not work there, crashed the DAW or produced other errors.  But it does work in OnParamChange.

I have built a half dozen plugins using this approach.  They work in VST, VST3, AU and AAX on both Windows and MacOS.  The only issue I've ever had was with Pro Tools as posted in this thread - and I have shared here how I worked that one out.


>... or to consider this a developer issue

I don't see how GetUI returning a valid pointer to an incomplete UI is a developer issue.  IMO, it's a framework issue.  In Pro Tools it throws an error.  Other DAWs somehow seem to work past it.  Perhaps Pro Tools is more finicky or perhaps the other formats have different construction ordering behind the scenes, IDK.  I do know for a fact that not all DAWs construct plugins in the same order or handle errors in the same way.

I have done a lot of homework on this and presented a lot of info here to support what I'm claiming. I posted this bug report to raise what I see as a problem along with a suggested solution/improvement.  I know other developers don't use OnParamChangeUI - they update everything in OnIdle - or don't build AAX- so maybe this issue just hasn't been widely experienced yet.

Anyhow, I have what works for me.  Please close this report or investigate at will.  Thank you.

**Comment by AlexHarker** [comment]
@rexmartin158 - I think there's probably quite a bit of frustration at both ends on this and I suspect that longer posts here won't resolve this so perhaps I can try to lay out some general principles of how I'm approaching this and suggest a way forward?

1 - I absolutely believe and accept that you are seeing an issue in your plugin in ProTools and that this could indicate a wider iplug2 issue, but it also could reflect a non-optimal design (or both) - I can't fully disentangle this yet.
2 - I can't reproduce on my setups, so I cannot be 100% sure why this is happening - I have not checked on ProTools as I am not set up to do so and getting set up to do so will take some time (so at least initially it seemed quicker to ask you to retest)
3 - I would really like to resolve this problem whether it be in your code or Iplug2 and appreciate the time you've put into looking at this.
4 - There seem to be some misunderstandings about how things should work in iplug2 as well as some of the things I've said, possibly to do with shorthands in writing long responses (so yes OnParamChangeUI() will be called in all formats before the UI is constructed, but that should be before it is constructed at all, not halfway through as you are seeing).
5 - I worked on this part of the UI code quite a bit with Oli so I have a very good idea of the details of what should and shouldn't be happening. I cannot see anywhere that my assumptions are broken in iplug2 (although I cannot run to check on AAX as the moment). I've tried to explain some of the details above of that, but I'm not sure that all of those explanations are clear enough - that's understandable, but it's not just a case of me denying the existence of any issue (in fact I outlined the main two reasons I think that GetUI() would return incorrectly above, but rather than just stopping this happening, I'd like to understand why it is the case in your plugin so the more fundamental issue can be resolved).

There are some things you are saying about your experience (OnParamChangeUI() not getting called for instance) that should not be the case. I don't know why that is, but until I fully understand the problem I don't want to implement changes that might be missing the underlying issue. There is also some interaction between your plugin code that seems highly relevant in this case.  

There are two ways forward I can see that might hopefully be less frustrating and lead to a fix.

A - if you can supply a project that is shareable and simple that demonstrates the issue I can investigate independently (whether or not I have to install Pro Tools and sort an AAX SDK to do so). IT doesn't need to be publicly shareable, but I'd need to be able to compile it.

B - As I suspect A might not be viable (or might be too time consuming) I'd be more than happy to try to find a time to work through this on a video call if you are able to do so. That would hopefully lead to a quicker diagnosis of the issue, and I'd be able to discuss some of the plugin things you are describing (such as manually calling OnParamChangeUI or related methods) and hopefully advise on a more iplug2 idiomatic approach (I'd also prefer to leave the interaction on a much more positive note, which I hope this would allow)..

Sorry we haven't managed to resolve this sooner. Thanks also for the clarification that you only see issues on AAX and it affects multiple plugins - that is very helpful (it doesn't fully rule out the issues of design, but it means that we are expecting  to find something in the AAX call chain that leads to the bug). Also I've missed commenting on this detail before (as there's already enough to discuss) but if your plugs work in other formats it is almost impossible that the problem occurs but it is ignored. This is because the issue will lead to dereferencing a null pointer, which will always crash, unless caught in an exception handler - there are no format wide exception handlers and I've never seen any of the big plugin hosts guarding against this, so we can have a high confidence that this is to do with AAX specific ordering.

**Comment by rexmartin158** [comment]
@AlexHarker 

OK, instead of trying to explain more in words I have built a simple example for you based on IPlugEffect.  I added a simple bitmap "control with tag" (an "LED" meter) that is updated in OnParamChangeUI that I call at the end of OnParamChange.

This plugin works in Reaper (as VST3) but does **_not_** work in Pro Tools - where it throws the same error as the title of this thread. 

Your comment that "> OnParamChangeUI() will currently always be called before the Layout function is called" doesn't make sense to me.  If I comment out my call to OnParamChangeUI the control does not update.  In other words, OnParamChangeUI is not automatically called by OnParamChange or anything else.

![Screeshot of AAE error on Pro Tools](https://user-images.githubusercontent.com/84396036/145687828-8e31bde1-fb16-4f7c-8166-0db200bfe245.png)

The cause of the error is a NULL pointer to the control that has not been created yet.  The attempt to update this control with GetUI()->GetControlWithTag(kCtrlTagMeter)->SetValueFromDelegate(gain); runs after the "if(GetUI()" check which means if(GetUI()) is returning "true" - which is wrong!  The control with tag does not yet exist.

![Screeshot of NULL pointer to control](https://user-images.githubusercontent.com/84396036/145687833-e4656e25-92c0-4483-a73b-cddefb11e71e.png)

Here is the call stack from Pro Tools debug:
UITest.aaxplugin!UITest::OnParamChangeUI(int paramIdx) Line 66	C++
UITest.aaxplugin!UITest::OnParamChange(int paramIdx) Line 56	C++
UITest.aaxplugin!iplug::IEditorDelegate::OnParamChange(int paramIdx, iplug::EParamSource source, int sampleOffset) Line 113	C++
UITest.aaxplugin!iplug::IEditorDelegate::OnParamReset(iplug::EParamSource source) Line 133	C++
UITest.aaxplugin!iplug::IPluginBase::UnserializeParams(const iplug::IByteChunk & chunk, int startPos) Line 136	C++
UITest.aaxplugin!iplug::IPluginBase::UnserializeState(const iplug::IByteChunk & chunk, int startPos) Line 146	C++
UITest.aaxplugin!iplug::IPlugAAX::SetChunk(unsigned int chunkID, const AAX_SPlugInChunk * pChunk) Line 504	C++
[External Code]

You may download the demo plugin project files from this link:  [GetUI()->GetControlWithTag(kCtrlTagMeter)->SetValueFromDelegate(gain);](url)

Again, based on the order of events I see in this plugin construction I have no idea how this works in VST3 or anywhere else.  OnParamChangeUI should not be allowed to run until mLayoutFunc is complete.

**Comment by AlexHarker** [comment]
Thanks @rexmartin158 - this is all very useful and I hope will allow us to resolve this soon.

There are a couple of potential issues I see in this code - the most pressing of which is that your definition of OnParamChangeUI() does not override a base method. Whereas there are two versions of OnParamChange() (with one or two arguments) there is only one of OnParamChangeUI() (with two arguments). This is a detail above in the discussion I did not pick up on. Therefore your method OnParamChangeUI() will never be called unless you call it manually.. If you define it with two parameters then it gets called (in the app version at a least) as expected.

The second possibly smaller issue is that your update to the GUI in OnParamChangeUI() depends on a class variable that you are updating in OnParamChange() - I would personally try to avoid that kind of dependency (and just redo getting the param value and dividing) so that the paths are independent of each other.

So, for this case(at least)  I can resolve the need for the manual call entirely. OnParamChangeUI() should be guaranteed to be called in the main thread only (whereas OnParamChange() may be called in the audio thread) so that's one reason it's not considered good practice to call OnParamChangeUI() from there.

For a simple case like this there is also no need for a tag - both controls could be bound to the kGain parameter. In cases where a control is based on multiple parameters then the best approach is a custom(used) control that references all the relevant parameters. I suspect that might be the kind of situation you are involved in, but I can't directly advise on that without understanding what the controls in your plugin should do (an example of that is the EQ/freqency control here, which references all the EQ parameters, and a few other things - over 10 parameters - https://www.surrealmachines.com/impact-demo/ )

**Comment by rexmartin158** [comment]
Thank you for the tips and info however I provided this project just to demonstrate the issue that the safety check "if(GetUI())" sometimes returns "true" when in fact the UI doesn't exist.  That, IMO, is a bug.  Even if OnParamChangeUI is called from the wrong location - or at the wrong time - that safety check should stop UI code within OnParamChangeUI from running - but it doesn't.

> There are a couple of potential issues I see in this code - the most pressing of which is that your definition of OnParamChangeUI() does not override a base method. Whereas there are two versions of OnParamChange() (with one or two arguments) there is only one of OnParamChangeUI() (with two arguments). This is a detail above in the discussion I did not pick up on. Therefore your method OnParamChangeUI() will never be called unless you call it manually.. If you define it with two parameters then it gets called (in the app version at a least) as expected.

None of the iPlug example projects - not even IPlugControls - use OnParamChangeUI so I tried to figure it out from reading elsewhere.  Is there an official example of how to use OnParamChangeUI - and when/for what purpose?

> The second possibly smaller issue is that your update to the GUI in OnParamChangeUI() depends on a class variable that you are updating in OnParamChange()

I'm confused by this comment.  Isn't that the whole purpose of OnParamChangeUI?

If I have done many systemically wrong things in my code why is it only producing an error in Pro Tools?  It works perfectly everywhere else (Logic, Cubase, Reaper, Cakewalk, Ableton, etc.).

For example, if calling "SendCurrentParamValuesFromDelegate()" from OnParamChange is a bad idea (yet the only place I have found it works) where is the intended/correct place to call it?

> an example of that is the EQ/freqency control here, which references all the EQ parameters, and a few other things - over 10 parameters - https://www.surrealmachines.com/impact-demo/

The more I work with iPlug the less I seem to understand it.  I will look this over.  Thank you for the example.

**Comment by AlexHarker** [comment]
Hi @rexmartin158  I'm aware that GetUI() may return incorrectly if it is called during or concurrently with UI construction. This could likely be fixed (I pretty much know how to do this, but I haven't coded it), although it would need to be done carefully to ensure other errors don't occur as a result. However, at present I still don't know *why* it's getting called at the wrong time (whether this is a threading error or an ordering error), because the stack trace above doesn't indicate that we are in the middle of the UI being created. The why (to my mind) might point to a much more important problem that GetUI() returning incorrectly, and if so, that's the one that needs fixing.

I was not aware that none of the examples show you how to use OnParamChangeUI() - we should definitely address that. I'm not sure on docs - I'll try to look into that also. In the meantime here's the simplest public code I have that uses it:

https://github.com/AlexHarker/HISSTools_Granular/blob/master/HISSToolsGranular.cpp

If you explicitly mark your methods as overrides you may also get warnings from your compiler for cases like this where the signatures don't match.

> I'm confused by this comment. Isn't that the whole purpose of OnParamChangeUI?

Not exactly - the purpose is to update the UI based on a parameter changing.

> If I have done many systemically wrong things in my code why is it only producing an error in Pro Tools? It works perfectly everywhere else (Logic, Cubase, Reaper, Cakewalk, Ableton, etc.).

That's the bit I'm most interested in - it means that Pro Tools is interacting with iplug in a different way to the other hosts/formats and that suggests that our assumptions about what happens in what order/on what thread might be wrong.

> if calling "SendCurrentParamValuesFromDelegate()" from OnParamChange is a bad idea (yet the only place I have found it works) where is the intended/correct place to call it?

It is fairly unlikely that a plugin developer should ever need to call it manually. I checked my plugins and I never call it as far as I can see. It gets called when the UI opens and in some other places by the framework.

**Comment by rexmartin158** [comment]
> It is fairly unlikely that a plugin developer should ever need to call it manually. I checked my plugins and I never call it as far as I can see. It gets called when the UI opens and in some other places by the framework.

OK, then what is the means to update/refresh controls with parameters when the parameter values have been changed by something in code, for example via SendParameterValueFromUI?  How do you update/refresh the UI so it shows the new parameter value?

I'm doing what I'm doing because I saw other developers ask this same question on the slack channel, etc., and were told to follow SendParameterValueFromUI with SendCurrentParamValuesFromDelegate() and - except in Pro Tools on initial construction - it works.  So if that's not right, what is the right way?

> I was not aware that none of the examples show you how to use OnParamChangeUI() - we should definitely address that. I'm not sure on docs - I'll try to look into that also. In the meantime here's the simplest public code I have that uses it:
> 
> https://github.com/AlexHarker/HISSTools_Granular/blob/master/HISSToolsGranular.cpp

Your code in this example looks considerably different than the IPlug2 example projects that I have been using as templates.  What are all these "Mutex locks" you have in various places?  And why do you have your own LayoutUI function?  Is this iPlug1 or iPlug2 code?  My setup/layout code looks like this:

```
  mMakeGraphicsFunc = [&]()
  {
    return MakeGraphics(*this, PLUG_WIDTH, PLUG_HEIGHT, PLUG_FPS, GetScaleForScreen(PLUG_WIDTH, PLUG_HEIGHT));
  };

  mLayoutFunc = [&](IGraphics* pGraphics)
  {
    pGraphics->AttachBackground(BACKGROUND_FN);
    pGraphics->LoadFont("Poppins-Medium", POPPINS_FN);
    pGraphics->AttachCornerResizer(EUIResizerMode::Scale, false, IColor(COLOR_WHITE), IColor(COLOR_WHITE), IColor(COLOR_WHITE), 17.f);
...
```

**Comment by AlexHarker** [comment]
> when the parameter values have been changed by something in code, for example via SendParameterValueFromUI

It depends what code is change the parameter and why- I probably also never call SendParameterValueFromUI manually (it is called in IControl::SetDirty(). So, for most purposes I'd simply link controls to parameters and use SetDirty() within the control - if there's a reason that doesn't work for you then if you tell me what you need to achieve I can try to advise on that.

> What are all these "Mutex locks" you have in various places? And why do you have your own LayoutUI function? Is this iPlug1 or iPlug2 code? My setup/layout code looks like this:

The mutex locks make sure that my core DSP class is updated in a thread safe manner. They used to exist in IPlug1 and in my case I think I need them to be sure things are right, so I use them manually anywhere that I'm accessing that class.

The LayoutUI() override is simply an alternative to using mLayoutFunc - I prefer the look of it, but it's functionally equivalent - just you write the layout as a method, rather than assign it to a variable during your constructor.

**Comment by AlexHarker** [comment]
Sorry - I was incorrect - I actually use SendParameterValueFromUI() in the code above, but at that point the UI has updated, so I don't need to send anything back to the UI. In any case it's also worth noting that SendParameterValueFromUI() will call through to OnParamChangeUI() (all of this is bit hard to trace through - which is another reason it is best avoided). 

However, I'm not 100% sure that all my calls are correct in that code - The calls to SendParameterValueFromAPI() look redundant - I'll need to look into that...

**Comment by rexmartin158** [comment]
>In any case it's also worth noting that SendParameterValueFromUI() will call through to OnParamChangeUI() 

OK, except OnParamChangeUI isn't used to update controls that are linked to parameters (AFAIK).  How do you update controls that are linked to parameters?  For example we could change the value of a parameter using SetValue() but all that does is change the parameter - it doesn't update the UI or automation data, etc.  How do we "refresh" everything else to match?

> all of this is bit hard to trace through - which is another reason it is best avoided unless really needed

I agree! I have a hard time tracing through much of iPlug - mostly because I don't understand all the "under the hood" stuff.  One function calls another function which calls another function linked to some other function that loops back to the second function, etc. - and it sometimes seems to go off on tangents having nothing to do with the original task.  Maybe it's just me, IDK, but I get very tangled up.  Nevertheless, it's still better than trying to write 7 different versions of each plugin directly from the SDKs so I greatly appreciate it!

**Comment by AlexHarker** [comment]
Thanks for this clarity - it is now much clearer why you need to update the UI in this way. I need to think about this a bit longer to figure out how I'd do this. If you only had buttons (with no parameters attached) to set these presets it would be easier, but because those presets can be set by parameter that's a bit more complicated. I presume that's a desired part of the design (or something you can no longer change if your plugin has been released)?

**Comment by rexmartin158** [comment]
> I presume that's a desired part of the design (or something you can no longer change if your plugin has been released)?

Yes, and like I've said it works!  I've used this same approach to update other interactive controls in other plugins that also work.

> all of this is bit hard to trace through

Yes, and now I am confused again.  Why is the function "SendCurrentParamValuesFromDelegate" named that?  According to its function code all that function does is loop through all the parameters calling "SendParameterValueFromDelegate" which does nothing else (unless overridden) but call OnParamChangeUI.  There is nothing in any of those functions that sends or sets parameter values.  What it _does_ do - somehow - is update all UI controls attached to parameters - not just controls set in OnParamChangeUI.  How is it doing _that_?  Is there more code to this function buried somewhere that I'm not seeing?

Yes.  There is a lot here I don't understand...

**Comment by AlexHarker** [comment]
In the current implementation when you click the button and you are recording automation in your DAW do you get parameter updates in the host for all changed parameters?

**Comment by olilarkin** [comment]
Unfortunately I don't have time to read all this. I finally got my ProTools license renewed and I downloaded the example project and removed the call to OnParamChangeUI() in OnParamChange() (that is wrong if not mentioned already). I input some automation, hit play and rapidly close and open the UI and I get an access violation eventually with no obvious issue in the call stack. Did this on windows.

There is something wrong here but I am not sure it is exactly as described

**Comment by rexmartin158** [comment]
@olilarkin 
> There is something wrong here but I am not sure it is exactly as described, since I can't get the -14034 error, once i remove the incorrect call to OnParamChangeUI()

The root issue of my report here is that GetUI() - at least in Pro Tools - sometimes returns a "valid" pointer to a UI that doesn't exist.  

As shown in the debugger screenshot below, the attempt to update a control with GetUI()->GetControlWithTag()->SetValueFromDelegate() tries to run even though the safety check "if(GetUI()" should be returning "false".

![image](https://user-images.githubusercontent.com/84396036/146276898-2fcc7950-98f1-4bbd-8c2d-c684565dc80d.png)

This is what throws the AAE error in Pro Tools.  For some reason this does not cause a problem in any other DAW - although it seems it should.

**Comment by AlexHarker** [comment]
Hi @olilarkin - it's great that you are set up for testing in ProTools - I'm still thinking about the correct solution to what @rexmartin158 is trying to do (which I think probably requires sending parameters to the GUI as if from the API, to make sure they get sent on the main thread), but knowing the answer to the question I posted above would help to understand the situation fully.

What would be super helpful at this point in terms of the larger possible issue is to know which thread crashes with the code with the OnParamChangeUI() included - if that's the main thread is it obvious that we are mid UI layout, or if not the main thread, is the main thread at that point?

**Comment by sstillwell** [comment]
I think maybe part of the issue is misinterpreting what GetUI() returns.  My assumption is that its only responsibility is to return a pointer to an initialized IGraphics context...and that doesn't say anything about whether all the controls are created or not - just that IGraphics is there ready to do your bidding.

I might be wrong, but it's a defensive, safe-ish assumption.

**Comment by rexmartin158** [comment]
>My assumption is that its only responsibility is to return a pointer to an initialized IGraphics context...and that doesn't say anything about whether all the controls are created or not - just that IGraphics is there ready to do your bidding.

That is what it seems to be doing however all the examples I see of updating graphics elements (controls) always preface the update with the check "if (GetUI())".  So if that does not assure the existence of the UI element what is the purpose of that check?

**Comment by sstillwell** [comment]
> That is what it seems to be doing however all the examples I see of updating graphics elements (controls) always preface the update with the check "if (GetUI())". If that check does not assure the existence of the UI element being updated what is the purpose of it in that case?

Without a valid IGraphics pointer it's all going to fail, isn't it?  So from that standpoint the check makes sense.  I think it may be slightly a legacy IPlug-ism.  Anyway, I won't guess further, I'll just be digging myself a deeper hole.

**Comment by AlexHarker** [comment]
The GetUI() check is necessary in certain places because the UI may or may not exist - as plugin devs may call those places themselves there is no easy universal place to put these checks (which would be another design design). Using GetUI() is (ASFAIK) an iPlug2 only thing, as in iPlug 1 the GUI did not get deleted/created with the window.

> I think maybe part of the issue is misinterpreting what GetUI() returns. My assumption is that its only responsibility is to return a pointer to an initialized IGraphics context...and that doesn't say anything about whether all the controls are created or not - just that IGraphics is there ready to do your bidding.

At the moment this is a correct interpretation, but it's a different question whether this is the correct design. However, the *assumptions* are that the framework will only ever call UI methods on the main thread and that UI methods will never be called during UI layout. If these are true that GetUI() should only return true once the UI is fully created. If not the behaviour is likely undefined/liable to crash at present.

In all the example projects this is in formats other than AAX certainly true. It may or may not be true for AAX, although the manual call to OnParamChangeUI() in the provided project complicates debugging this issue. Now that @olilarkin is set up to test I would be able to take the info in this thread to work with him to answer this (if we can both find some time for it).

**Comment by sstillwell** [comment]
> sstillwell - your issues with crashes in OnIdle() may be different from what rexmartin158 is experiencing. A reduced test project exhibiting this issue would still be useful if you are able to produce one.

I tried to reproduce it and couldn't...it may be an artifact from an earlier version of IPlug2 that I just never removed.  Pay no attention to the man behind the curtain.

**Comment by rexmartin158** [comment]
>At the moment this is technically a correct interpretation, but it's a different question whether this is the correct design and what to expect in practice. The _assumptions_ are that the framework will only ever call UI methods on the main thread and that UI methods will never be called during UI layout. If these are true that GetUI() should only return true once the UI is fully created. If not the behaviour is likely undefined/liable to crash at present.

Maybe it's just my interpretation but the name "GetUI" - to me - implies the whole UI.  In that case, the check if(GetUI()) could be made more robust - and not rely on assumed timing of events - if it checked to make sure mLayoutFunc is completed before setting its pointer valid.  That's essentially what my hack does.

**Comment by asedev** [comment]
Let me throw in my two cents.
1. There is a good reason to have a "GetUI()" function which simply returns a context to the UI in general.
2. Whether the UI is already in a "useable" state is nothing that could be implemented "generally".
3. We could add some generalizations like "bool GetUIReady() {return true;}" which finally has to be implemented by each developer/plugin itself.

**Comment by sstillwell** [comment]
> 3. We could add some generalizations like "bool GetUIReady() {return true;}" which finally has to be implemented by each developer/plugin itself.

Or `bool GetUIReady() {return (GetUI() && true);}` as the default "Please override me" implementation so you don't have to call both in your code.

I think that's a good addition to the code @asedev.

**Comment by rexmartin158** [comment]
> 1. There is a good reason to have a "GetUI()" function which simply returns a context to the UI in general.

What is that reason?  What would you be doing, requiring checking if(GetUI()), if not updating something in the UI?

**Comment by asedev** [comment]
For example:
GetUI() -> you may start adding controls
GetUIReady() -> you may start updating controls

**Comment by AlexHarker** [comment]
Thanks all for contributions - at this point I think the best way forward is that Oli and I pic this up as we can (probably in the New Year) bearing in mind the suggestions and information here, after we've had the chance to do some more testing.

For clarity GetUI() is intended to give you the UI pointer when it exists - that has always been understood to mean after creation but the implementation is actually just once IGraphics is made. Designs in which there is a call to GetUI() between that point and the end of layout have not been considered (as this isn't the design for any examples or plugs we have) but this thread makes it totally clear what the issues and possibly design choices are that affect this issue, so from that we can make some decisions on how to approach this. There may also be some hidden bug(s) in AAX to look at, but we now have enough info to test and one of us has the ability to do so. Thanks all for bearing with and contributing and I'll update when there's something more concrete.

**Comment by AlexHarker** [comment]
A quick update - I've just gone through this with @olilarkin and we now understand the issue which matches one of my hunches above (that threading assumptions were being broken)

Pro Tools is trying to unserialise the state on a thread other than the main thread. This doesn't happen when the plugin is constructed anew, but it will happen on project reload, or if you delete a plugin and then undo (and time the plugin is constructed with a known state).

The best fix is probably to:

- check if we are on the main thread
- if so unserialise
- if not copy the chunk and defer unserialising until the next timer call

That's possible, but not totally trivial - I'll try to code something up when I can for Oli to test.
Thanks for reporting this @rexmartin158 and your patience - this is a significant bug in AAX and it has possible implications beyond the crash above and needs fixing.

**Comment by rexmartin158** [comment]
@AlexHarker - thank you for looking into this.  Yes, my little workaround has been solving the problem for now.  I haven't seen any ill side effects from it so I'm pressing on.  The proper fix will certainly be better for everyone than my "band-aid".  Everything else in AAX seems to be working OK.

**Comment by olilarkin** [comment]
There is an AAX Flag "AAX_eProperty_RequiresChunkCallsOnMainThread" that we should probably set

**Comment by olilarkin** [comment]
I didn't realise it was deprecated. As it stands, I think we probably still have to do some work to make AAX state save/recall threadsafe

If you'd like to submit a PR addressing this problem, we'd love to have your contributions!

**Comment by sstillwell** [comment]
Not really my area of expertise, but I'm digging in to see what I can do.  I've got a working build, at any rate, but whether or not it does what we want is a completely different thing. :)
